From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 17 May 2024 09:46:45 +0300
Subject: [PATCH] Init/destruct injectors

System property in tests is to bandaid failing tests due to missing "--add-opens java.base/java.lang=ALL-UNNAMED" flag in gradle build task required by Material injector patch.

diff --git a/src/main/java/net/minecraft/server/Bootstrap.java b/src/main/java/net/minecraft/server/Bootstrap.java
index 8c2f9f549b5155b8d8fecbc22164b334805381e2..fb3525ed6c34f3b87063c8e5d4857704ccd5d8ef 100644
--- a/src/main/java/net/minecraft/server/Bootstrap.java
+++ b/src/main/java/net/minecraft/server/Bootstrap.java
@@ -62,6 +62,12 @@ public class Bootstrap {
             Bootstrap.isBootstrapped = true;
             Instant instant = Instant.now();
 
+            // Kiterino start - Init/destruct injectors
+            if (!Boolean.getBoolean("Kiterino.testingBase")) {
+                me.sosedik.kiterino.util.KiterinoDataInjectorImpl.init(); // Kiterino - Real custom items
+                me.sosedik.kiterino.util.KiterinoBootstrapMaterialInjectorImpl.init(); // Kiterino - Injecting custom Materials
+            }
+            // Kiterino end - Init/destruct injectors
             io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enterBootstrappers(); // Paper - Entrypoint for bootstrapping
             if (BuiltInRegistries.REGISTRY.keySet().isEmpty()) {
                 throw new IllegalStateException("Unable to load registries");
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 4c3e1b4d8f972a3773480eabaf549aa52c2f4ab2..3e2cec5acbaa4942487a38fd0f5679bd0c0b480b 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -333,9 +333,20 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
         // Purpur end
 
+	    // Kiterino start - Init/destruct injectors
+	    boolean skipInject = Boolean.getBoolean("Kiterino.testingBase");
+	    var materialInjectorK = skipInject ? null : me.sosedik.kiterino.util.KiterinoBootstrapMaterialInjectorImpl.destruct();
+	    // Kiterino end - Init/destruct injectors
         // CraftBukkit start
         // this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage)); // Spigot - moved up
         this.server.loadPlugins();
+        // Kiterino start - Init/destruct injectors
+        if (materialInjectorK != null) {
+            // Delaying nms inject up until onLoad() were called to allow passing custom data not available in bootstrap into constructors
+            materialInjectorK.finalizeInject(me.sosedik.kiterino.util.KiterinoDataInjectorImpl.getInstance());
+            me.sosedik.kiterino.util.KiterinoDataInjectorImpl.destruct();
+        }
+        // Kiterino end - Init/destruct injectors
         this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
         // CraftBukkit end
 
diff --git a/src/test/java/org/bukkit/support/AbstractTestingBase.java b/src/test/java/org/bukkit/support/AbstractTestingBase.java
index 4e9612600822bc1b52b8528dff2783d43385ec73..f597e8144e1ac7bea55df12bd421f3a94c747084 100644
--- a/src/test/java/org/bukkit/support/AbstractTestingBase.java
+++ b/src/test/java/org/bukkit/support/AbstractTestingBase.java
@@ -44,6 +44,7 @@ public abstract class AbstractTestingBase {
 
     static {
         System.setProperty("Paper.pushPaperAssetsRoot", "true"); // Paper
+        System.setProperty("Kiterino.testingBase", "true"); // Kiterino - Init/destruct injectors
         SharedConstants.tryDetectVersion();
         Bootstrap.bootStrap();
         // Populate available packs
